acme/acme.go

```diff
@@ -0,0 +1,404 @@
+package acme
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	"github.com/xenolf/lego/acme"
+	"io/ioutil"
+	fmtlog "log"
+	"os"
+	"reflect"
+	"sync"
+	"time"
+)
+
+// Account is used to store lets encrypt registration info
+type Account struct {
+	Email              string
+	Registration       *acme.RegistrationResource
+	PrivateKey         []byte
+	DomainsCertificate DomainsCertificates
+}
+
+// DomainsCertificates stores a certificate for multiple domains
+type DomainsCertificates struct {
+	Certs []*DomainsCertificate
+	lock  *sync.RWMutex
+}
+
+func (dc *DomainsCertificates) init() error {
+	if dc.lock == nil {
+		dc.lock = &sync.RWMutex{}
+	}
+	dc.lock.Lock()
+	defer dc.lock.Unlock()
+	for _, domainsCertificate := range dc.Certs {
+		tlsCert, err := tls.X509KeyPair(domainsCertificate.Certificate.Certificate, domainsCertificate.Certificate.PrivateKey)
+		if err != nil {
+			return err
+		}
+		domainsCertificate.tlsCert = &tlsCert
+	}
+	return nil
+}
+
+func (dc *DomainsCertificates) renewCertificates(acmeCert *Certificate, domain Domain) error {
+	dc.lock.Lock()
+	defer dc.lock.Unlock()
+
+	for _, domainsCertificate := range dc.Certs {
+		if reflect.DeepEqual(domain, domainsCertificate.Domains) {
+			domainsCertificate.Certificate = acmeCert
+			tlsCert, err := tls.X509KeyPair(acmeCert.Certificate, acmeCert.PrivateKey)
+			if err != nil {
+				return err
+			}
+			domainsCertificate.tlsCert = &tlsCert
+			return nil
+		}
+	}
+	return errors.New("Certificate to renew to found from domain " + domain.Main)
+}
+
+func (dc *DomainsCertificates) addCertificateForDomains(acmeCert *Certificate, domain Domain) (*DomainsCertificate, error) {
+	dc.lock.Lock()
+	defer dc.lock.Unlock()
+
+	tlsCert, err := tls.X509KeyPair(acmeCert.Certificate, acmeCert.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+	cert := DomainsCertificate{Domains: domain, Certificate: acmeCert, tlsCert: &tlsCert}
+	dc.Certs = append(dc.Certs, &cert)
+	return &cert, nil
+}
+
+func (dc *DomainsCertificates) getCertificateForDomain(domainToFind string) (*DomainsCertificate, bool) {
+	dc.lock.RLock()
+	defer dc.lock.RUnlock()
+	for _, domainsCertificate := range dc.Certs {
+		domains := append([]string{domainsCertificate.Domains.Main}, domainsCertificate.Domains.SANs...)
```

Comment:
`domainsCertificate.Domains.Main` is a slice right ? Then, `append` checks if the slice needs to grow or not. If it does need to grow, it will allocate a new array and `domainsCertificate.Domains.Main` is not modified (everything is copied/moved) to the new array. But if it does not need to grow, it will just add the element to the current array, and thus, `domainsCertificate.Domains.Main` is modified.

Something like that :

```go
a := []string{ "Janvier", "Fevrier", "Mars" }
b := append(months, "Avril")
c := append(months, "Mai")

fmt.Println(len(a), cap(a)) // "3, 3"
fmt.Printf("%q\n", a)       // ["Janvier" "Février" "Mars"]
fmt.Println(len(b), cap(b)) // "4 6"
fmt.Printf("%q\n", b)       // ["Janvier" "Février" "Mars" "Avril"]
fmt.Println(len(c), cap(c)) // "5 6"
fmt.Printf("%q\n", c)       // ["Janvier" "Février" "Mars" "Avril" "Mai"]

fmt.Printf("%q\n", b[:5])   // ["Janvier" "Février" "Mars" "Avril" "Mai"]
// … mais pas a
fmt.Printf("%q\n", a[:5])   // panic : out of range
```

I think it would be better create a new slice and put the elements of both inside it :angel:.